pkg/receive/multitsdb.go

```diff
@@ -0,0 +1,269 @@
+	terrors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/component"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/shipper"
+	"github.com/thanos-io/thanos/pkg/store"
+	"golang.org/x/sync/errgroup"
+)
+
+type MultiTSDB struct {
+	dataDir         string
+	logger          log.Logger
+	reg             prometheus.Registerer
+	tsdbCfg         *tsdb.Options
+	tenantLabelName string
+	labels          labels.Labels
+	bucket          objstore.Bucket
+	upload          bool
+
+	mtx         *sync.RWMutex
+	dbs         map[string]*FlushableStorage
+	appendables map[string]*tsdb.ReadyStorage
+	stores      map[string]*store.TSDBStore
+	shippers    map[string]*shipper.Shipper
+}
+
+func NewMultiTSDB(dataDir string, l log.Logger, reg prometheus.Registerer, tsdbCfg *tsdb.Options, labels labels.Labels, tenantLabelName string, bucket objstore.Bucket) *MultiTSDB {
+	if l == nil {
+		l = log.NewNopLogger()
+	}
+
+	return &MultiTSDB{
+		dataDir:         dataDir,
+		logger:          l,
+		reg:             reg,
+		tsdbCfg:         tsdbCfg,
+		mtx:             &sync.RWMutex{},
+		dbs:             map[string]*FlushableStorage{},
+		stores:          map[string]*store.TSDBStore{},
+		appendables:     map[string]*tsdb.ReadyStorage{},
+		shippers:        map[string]*shipper.Shipper{},
+		labels:          labels,
+		tenantLabelName: tenantLabelName,
+		bucket:          bucket,
+		upload:          bucket != nil,
+	}
+}
+
+func (t *MultiTSDB) Open() error {
+	if err := os.MkdirAll(t.dataDir, 0777); err != nil {
+		return err
+	}
+
+	return t.openTSDBs()
+}
+
+func (t *MultiTSDB) Close() error {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	errmtx := &sync.Mutex{}
+	merr := terrors.MultiError{}
+	wg := &sync.WaitGroup{}
+	for _, tsdb := range t.dbs {
+		tsdb := tsdb
+		wg.Add(1)
+		go func() {
+			if err := tsdb.Close(); err != nil {
+				errmtx.Lock()
+				merr.Add(err)
+				errmtx.Unlock()
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return merr.Err()
+}
+
+func (t *MultiTSDB) Flush() error {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	errmtx := &sync.Mutex{}
+	merr := terrors.MultiError{}
+	wg := &sync.WaitGroup{}
+	for _, tsdb := range t.dbs {
+		tsdb := tsdb
+		wg.Add(1)
+		go func() {
+			if err := tsdb.Flush(); err != nil {
+				errmtx.Lock()
+				merr.Add(err)
+				errmtx.Unlock()
+			}
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return merr.Err()
+}
+
+func (t *MultiTSDB) Upload(ctx context.Context) error {
+	if !t.upload {
+		return nil
+	}
+
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	errmtx := &sync.Mutex{}
+	merr := terrors.MultiError{}
+	wg := &sync.WaitGroup{}
+	for tenant, s := range t.shippers {
+		level.Debug(t.logger).Log("msg", "uploading block for tenant", "tenant", tenant)
+		s := s
+		wg.Add(1)
+		go func() {
+			if uploaded, err := s.Sync(ctx); err != nil {
+				errmtx.Lock()
+				merr.Add(fmt.Errorf("failed to upload %d: %w", uploaded, err))
```

Comment:
```suggestion
				merr.Add(errors.Wrap(err, "sync %d", uploaded))
```

As per https://github.com/thanos-io/thanos/blob/8828b3b797ae1b75a19c0ccc4024b8c212c1adc9/docs/contributing/coding-style-guide.md#wrap-errors-for-more-context-dont-repeat-failed--there

Reply:
Done